pkg/storage/endpoint/gc_states.go

```diff
@@ -120,6 +120,79 @@ var (
 	_ json.Unmarshaler = (*ServiceSafePoint)(nil)
 )
 
+// TimeOptional is time.Time or nil
+type TimeOptional struct {
+	*time.Time
+}
+
+func (t *TimeOptional) int64() int64 {
+	if t.Time != nil {
+		return t.UnixNano()
+	}
+	return 0
+}
+
+// MarshalJSON implements json.Marshaler for TimeOptional
+func (t *TimeOptional) MarshalJSON() ([]byte, error) {
+	return json.Marshal(t.int64())
+}
+
+// UnmarshalJSON implements json.Marshaler for TimeOptional
+func (t *TimeOptional) UnmarshalJSON(b []byte) error {
+	var expireAt int64
+	if err := json.Unmarshal(b, &expireAt); err != nil {
+		return err
+	}
+	if expireAt < math.MaxInt64 && expireAt > 0 {
+		val := time.Unix(0, expireAt)
+		t.Time = &val
+	} else {
+		t.Time = nil
+	}
+	return nil
+}
+
+// GlobalGCBarrier represents a global GC barrier.
+// It looks like a GCBarrier now but the struct might change with the code evolve.
+// A more important reason is to distinguish from GC barrier by the type system,
+// avoiding potential misuse between them in code.
+type GlobalGCBarrier struct {
+	BarrierID string       `json:"barrier_id"`
+	BarrierTS uint64       `json:"barrier_ts"`
+	ExpiredAt TimeOptional `json:"expired_at"`
+}
+
+// NewGlobalGCBarrier creates a new GlobalGCBarrier. The given expirationTime will be rounded up to the next second if it's
+// not in integral seconds.
+// Passing nil to `expirationTime` means the barrier never expires.
+func NewGlobalGCBarrier(barrierID string, barrierTS uint64, expirationTime *time.Time) *GlobalGCBarrier {
+	// Round up the expirationTime.
+	if expirationTime != nil {
+		rounded := expirationTime.Add(time.Second - time.Nanosecond).Truncate(time.Second)
+		*expirationTime = rounded
+	}
```

Comment:
The rounding-up used in GC bariers is for handling the fact that the storage precision is in seconds level, and it doesn't make sense if you store it in nanoseconds precision.
However, please consider keep the rounding up and change the precision back to seconds.